proposals/pr002-hunkspace.md

```diff
@@ -0,0 +1,544 @@
+renders the localized resource to the user. The renderer of the
+message does not have access to the to-be rendered message ahead of
+time.
+
+Another instance of delegated localization is an application using
+platform-level localizations for common user interface elements ("OK",
+"Cancel", "Edit"\...).   Mechanisms to do this are reinvented in each
+independent piece of software.  A localization clearinghouse would allow
+the OS to centralize localization decisions.
+
+A way to resolve is to devise some form of a mini "rendering protocol"
+where the screen reader provides rendering hints, and the voice UI
+renderer follows the hints.  The claim here is that this specific
+pattern is common enough that localization could benefit from a common
+specification of that rendering protocol.
+
+# Detailed Proposal
+
+Allow independent third parties to put new non-overlapping points into
+a common hunkspace, such that for each platform a single data provider
+can be used to provide any localized resource.
+
+Specify a mechanism by which this can be done, e.g. by adopting an
+URI-like encoding of hunkspace points, and providing a two-way mapping
+between an URI and an "unpacked" version of a hunkspace key.
+
+**Why this complication?** It allows third party contributions to localized
+software in trusted computing environments.  For example, third party provision
+of localized resources, as described in the Fuchsia use case.
+
+**Why should ICU4X care?** Strictly speaking perhaps it doesn't need to. One
+can build a "hunkspace" data provider given a cldr data provider, and fixing
+the "source" dimension to be "org.unicode.cldr" (see below).  An advantage of a
+commonly understood specification is that it may motivate software authors to
+provide mutually compatible data provider interfaces.
+
+**Do we even need common schemas like this?** In a world where localizations
+are islands of functionality, maybe the answer is no.  A CLDR data provider can
+exist on its own.  Firefox has its own elaborate L10N framework.  Fuchsia OS
+has use cases where platform-level I18N/L10N data provision is needed, but
+these use cases can live inside Fuchsia indefinitely.  It starts to matter if
+these worlds get to interact at some level. For example Firefox L10N can not
+exist in its current extensible form on Fuchsia, due to the tight controls that
+the OS puts in place on the runtime. (You'd have to download a prebuilt version
+for your locale, from Mozilla.)  Having a cross-platform data provider
+specification could make that a non-issue.]
+
+## Unique IDs
+
+Every localizable resource should get a unique identifier.  This allows
+a platform-level localization clearinghouse (from the example above) to
+be a one-stop data provider for localization needs.
+
+For example, the [current proposal of the icu4x data provider for
+CLDR](https://github.com/unicode-org/icu4x/blob/master/components/data-provider/src/lib.rs%23L52) uses
```

Comment:
nit: broken link, encoding `#` when there's purpose as a fragment identifier breaks the URL (https://github.com/unicode-org/icu4x/blob/master/components/data-provider/src/lib.rs%23L52) (breaks on Chrome & Safari) vs. (https://github.com/unicode-org/icu4x/blob/master/components/data-provider/src/lib.rs#L52)

Reply:
Fixed.